File: nezal-ar/app/models/games/base_defender/building.rb
module BD
  class Building
    @@states = { 
               'NOT_PLACED' => 0,
               'UNDER_CONSTRUCTION' => 1,
               'UPGRADING' => 2,
               'NORMAL' => 3 
                }
    class << self

      def init
        { 'levels' => {} }
      end

      def states
        @@states
      end

      def neighbor_data
        []
      end

      def new_building_specs 
        {'level' => 0, 'coords' => {'x' => nil, 'y' => nil}}
      end

      def build(user_game_profile, coords, name=nil)
        building_name = name || @name
        game_data = Game::current.data
        location_hash = BaseDefender.convert_location(coords)
        
        validation = validate_building(user_game_profile, coords, building_name)
        return validation if validation['valid'] == false
        
        user_game_profile.idle_workers -= 1
        user_game_profile.data[building_name]= {} if user_game_profile.data[building_name].nil?
        user_game_profile.data[building_name][location_hash] = self.new_building_specs
        user_game_profile.data[building_name][location_hash]['startedBuildingAt'] = Time.now.utc.to_i
        user_game_profile.data[building_name][location_hash]['started_repairing_at'] = 0
        user_game_profile.data[building_name][location_hash]['state'] = states['UNDER_CONSTRUCTION']
        user_game_profile.data[building_name][location_hash]['coords'] = coords
        user_game_profile.data[building_name][location_hash]['hp'] = game_data['buildings'][building_name]['levels']['1']['hp']
        user_game_profile.rock -= game_data['buildings'][building_name]['levels']['1']['rock']
        user_game_profile.lumber -= game_data['buildings'][building_name]['levels']['1']['lumber']
        return validation
      end
      
      def validate_building(user_profile, coords, name=nil)
        game_metadata = Game::current.data
        building_name = name || @name
        #validating workers
        if( user_profile.idle_workers == 0 )
          return { 'valid' => false,
                   'error' => "There is no idle workers to build the " + building_name}
        end
        #validating resources
        neededRock = Game::current.buildings[building_name]['levels']['1']['rock'] - user_profile.rock;
        neededLumber = Game::current.buildings[building_name]['levels']['1']['lumber'] - user_profile.lumber;
        if( neededRock > 0 && neededLumber > 0 )
          return {'valid' => false,
                  'error' => "Not enough resources, you need more "+ neededRock +" rock and "+ neededLumber + " lumber"}
        end
        
        if( neededRock > 0)
          return {'valid' => false,
                  'error' => "Not enough resources, you need more " + neededRock + " rock"}
        end
        if( neededLumber > 0)
          return {'valid' => false,
                  'error' => "Not enough resources, you need more " + neededLumber + " rock"}
        end
        #validating dependencies
        dependencies_valid = validate_dependencies user_profile, building_name, '1'
        if(!dependencies_valid['valid'])
          return dependencies_valid
        end
        max_no_valid = validate_max_no_of_buildings user_profile, building_name, '1'
        if(!max_no_valid['valid'])
          return max_no_valid
        end
         #validating location # to be reconsidered  after server Map implementation
#        puts "user_profile_metadata['map'][coords['x']][coords['y']] : " + user_profile_metadata['map'][coords['x']][coords['y']].to_s
#        puts "BaseDefender.land_marks[@can_be_built_on] : " + BaseDefender.land_marks[@can_be_built_on].to_s
#        puts "coords['x'] : " + coords['x'].to_s
#        puts "coords['y'] : " + coords['y'].to_s
#        
#        if(user_profile_metadata['map'][coords['y']][coords['x']] != BaseDefender.land_marks[@can_be_built_on])
#          return {'valid' => false,
#                  'error' => @name + " can be built on "+ @can_be_built_on +" only!"}
#        end
        
        return {'valid' => true, 'error' => ''}
      end
      
      def validate_dependencies user_profile, building_name, level
        building_dependencies = Game::current.buildings[building_name]['levels'][level]['dependency']['buildings']
        building_dependencies.each do |key,value|
          found = false
          user_profile.data[key].values.each do |building|
            if(building['level'] >= value.to_i)
              found = true
              break
            end
          end
          if(!found)
            return {'valid' => false,
                  'error' => "Cannot build " + building_name + ", you need a "+ key.gsub('_',' ').capitalize + " level "+value }
          end
        end 
        return {'valid' => true, 'error'=>''}
      end
      
      def validate_max_no_of_buildings user_profile, building_name, level
        limiting_building = Game::current.buildings[building_name]['levels'][level]['limited_by']
        if(!limiting_building.nil?)
          max_level = 0 
          registery = user_profile.data[limiting_building]
          if(registery.nil?)
                return {'valid' => false,
                  'error' => "Cannot build " + building_name + ", you need a "+ limiting_building.gsub('_',' ').capitalize }
          end
          registery.each do |key, value|
            max_level = value['level'] if(value['level']>max_level)
          end
          if(max_level > 0)
            max_no_of_buildings = Game::current.buildings[limiting_building]['levels'][max_level.to_s]['limiting']['others'][building_name].to_i
            if(max_no_of_buildings.nil?)
              max_no_of_buildings = Game::current.buildings[limiting_building]['levels'][max_level.to_s]['limiting']['global'].to_i
            end
            if(!user_profile.data[building_name].nil? && max_no_of_buildings<=user_profile.data[building_name].size)
              return {'valid' => false,
                  'error' => "Cannot build more than #{user_profile_metadata[building_name].size} #{building_name}."}
            end
          else
             return {'valid' => false,
                  'error' => "You can only build " + max_no_of_buildings + " " + building_name.gsub('_',' ').capitalize}
          end
        end
        return {'valid' => true, 'error'=>''}
      end
      
      def move(user_game_profile, name, coords, old_coords)
        coords_str = BaseDefender.convert_location(coords)
        old_coords_str = BaseDefender.convert_location(old_coords)
        building = user_game_profile.data[name].delete(old_coords_str)
        if building.present?
          building['coords'] = coords
          user_game_profile.data[name][coords_str] = building
          return {'valid' => true, 'error' => ''}
        else
          return {'valid' => false, 'error' => "There is no #{name} in this location"}
        end
      end

      def upgrade(user_game_profile, coords, name=nil)
        building_name = name || @name
        location_hash = BaseDefender.convert_location(coords)
        game_data = Game::current.data
        validation = validate_upgrade(user_game_profile, coords, building_name)
        return validation if validation['valid'] == false      
        building = user_game_profile.data[building_name][location_hash]
        current_level = building['level']
        
        user_game_profile.idle_workers -= 1
        building['startedBuildingAt'] = Time.now.utc.to_i
        building['state'] = states['UPGRADING']
        building['coords'] = coords
        building_level_info = game_data['buildings'][building_name]['levels'][(current_level+1).to_s]
        building['hp'] = building_level_info['hp']      
        user_game_profile.rock -= building_level_info['rock']
        user_game_profile.lumber -= building_level_info['lumber']
        return validation
      end
      
      def validate_upgrade(user_profile, coords, name=nil)
        building_name = name || @name
        game_data = Game::current.data
        #validating workers
        if( user_profile.idle_workers == 0 )
          return { 'valid' => false,
                   'error' => "There is no idle workers to build the " + building_name}
        end
        
        #make sure the building we are attempting to upgrade exists
        location_hash = BaseDefender.convert_location(coords)
        if( !user_profile.data[building_name][location_hash] )
          return {'valid' => false,
                  'error' => "The building you are trying to upgrade doesn't exist"}
        end
        #validating there is a next level to upgrade too, validate the user has enough resources.
        current_level = user_profile.data[building_name][location_hash]['level']

        if( game_data['buildings'][building_name]['levels'][(current_level+1).to_s] )
          neededRock = game_data['buildings'][building_name]['levels'][(current_level+1).to_s]['rock'] - user_profile.rock;
          neededLumber = game_data['buildings'][building_name]['levels'][(current_level+1).to_s]['lumber'] - user_profile.lumber;
          if( neededRock > 0 && neededLumber > 0 )
            return {'valid' => false,
                    'error' => "Not enough resources, you need more #{neededRock} rock and #{neededLumber} lumber"}
          end
          if( neededRock > 0)
            return {'valid' => false,
                    'error' => "Not enough resources, you need more #{neededRock} rock"}
          end
          if( neededLumber > 0)
            return {'valid' => false,
                    'error' => "Not enough resources, you need more #{neededLumber} rock"}
          end
          # TODO : In here should come a part to check the dependency chain.
          # make sure that the user have met the conditions required for that upgrade.
        else
          return {'valid' => false,
                  'error' => "Max upgrade level reached"}
        end
        dependencies_valid = validate_dependencies user_profile, building_name, (current_level+1).to_s
        if(!dependencies_valid['valid'])
          return dependencies_valid
        end
        return {'valid' => true, 'error' => ''}
      end
      
    end
  end
end

File: nezal-ar/public/base-defender/js/game/building.js
var Building = Class.create({
  factory : null,
  level : null,
  remainingBuildTime : null,
  _LoadTime : null,
  coords : {x : null, y : null},
  currentLevelBluePrints : null,
	hp : 0,
	maxHp :0,
  nextLevelBluePrints : null,
	locationValid : true,
  state : 0,
	working : true,
	repairing : false,
	states : {NOT_PLACED : 0, UNDER_CONSTRUCTION : 1, UPGRADING : 2, NORMAL : 3},
  initialize : function(factory, buildingSpecs){
    this.name = factory.name;
    this.game = factory.game;
    this.canBeBuiltOn = factory.canBeBuiltOn;
    this.factory = factory;
    this._LoadTime = new Date().getTime();
    this.level = buildingSpecs.level;
    this.coords = buildingSpecs.coords;
    this.remainingBuildTime = buildingSpecs.remainingTime;
    this.currentLevelBluePrints = this.factory.bluePrints['levels'][this.level];
    this.nextLevelBluePrints = this.factory.bluePrints['levels'][this.level+1];
	if(this.level>0)this.maxHp = this.currentLevelBluePrints.hp
	else this.maxHp = this.nextLevelBluePrints.hp
	if(buildingSpecs.hp!=null)this.hp = buildingSpecs.hp;
	else{
		this.hp = this.maxHp;
	} 
	if(this.hp<this.maxHp ){
		this.working = false
		if(!buildingSpecs.started_repairing_at) this.game.attackManager.showRepairMsg()
	} 
	if(buildingSpecs.started_repairing_at > 0){
		this.startRepairing()
	}
    this.initialState = buildingSpecs.state;
    this.stateNotifications = {};
    for(var state in this.states){
      this.stateNotifications[this.states[state]] = [];  
    }
  },
  
	startRepairing : function(){
		this.repairing = true
		var self = this
		condition = function(){
			return self.hp >= self.maxHp
		}
		mainFunc = function(){
			if(!self.game.attackManager.attacking)
				self.hp+=40
		}	
		callback = function(){
			self.hp = self.maxHp
			self.repairing = false
			self.working = true
			self.setState(self.states.NORMAL)
		}
		var ticks = self.game.reactor.everySeconds(1)
		self.game.reactor.pushPeriodicalWithCondition(ticks , mainFunc, condition, callback)
	},
	
  init : function(){
    this.setState(this.initialState);
    return this;
  },
  
  tick : function(){
    var self = this;
    if (this.state == this.states.UNDER_CONSTRUCTION || this.state == this.states.UPGRADING ) {
      if(this.elapsedTime() >= this.nextLevelBluePrints.time) {
        var delayRequest = this.game.scene.reactor.everySeconds(2);
        self.game.scene.reactor.push(delayRequest, function(){
          self.game.reInitialize();
        });
      }
    } else if (this.state == this.states.NOT_PLACED) {
			this.locationValid = this.validateLocation(this.coords.x,this.coords.y);
	  }
  },
	
	validateLocation : function(x,y){
		var mapCoords = this.game.scene.map.tileValue(x,y);
		var map = this.game.scene.map;
		var cond1 = map.grid[mapCoords[0]][mapCoords[1]].terrainType == this.game.scene.landmarks.get(this.canBeBuiltOn);
		var cond2 = map.validateLocation(this)
    if(!cond1){
      Notification.alert(this.name + " can be built on " + this.canBeBuiltOn + " only!");
      return false;
    }
		if(!cond2){
      //console.log("this location is occupied by another building");
      return false;
    }
		return true;
	},
  
  stateChanged : function(){
  },
  
  setState : function(newState){
    this.state = newState;
    this.stateNotifications[newState].each(function(fn){fn()})
  },
  
  
  
  /** This should return if we need to off the building mood or no*/
  build : function(xBlock, yBlock){
    this.coords['x'] = xBlock;
    this.coords['y'] = yBlock;    
    if(this.isValidToBuild(xBlock, yBlock)) {
      var response = this.game.network.upgradeBuilding(this.name, this.coords);
	    Sounds.play(Sounds.gameSounds.constructions)
		this.game.buildingMode = null
      this.game.updateGameStatus(response['gameStatus']);
      return response['done'];
    } else {
      //this.game.buildingMode.cancelBuildingMode();
      return false;
    }
  },

  upgrade: function(){
    if(this.isValidToUpgrade()){
      var response = this.game.network.upgradeBuilding(this.name, this.coords);
      this.game.updateGameStatus(response['gameStatus']);
    }
  },
	
	move : function(x,y){
		this.coords['x'] = x;
	    this.coords['y'] = y;
	    if(this.validateLocation(x,y)){
		  this.setState(this.states.NORMAL)
	      var response = this.game.network.moveBuilding(this.name, this.coords, this.oldCoords);
	      this.game.updateGameStatus(response['gameStatus']);
	      return response['done'];
	    }else{
		  //this.game.reInitialize()
	      return false;
	    }
	},
  
	inProgress : function(){
    	return this.state == this.states.UNDER_CONSTRUCTION || this.state == this.states.UPGRADING;		
	},
	
  elapsedTime : function(){
    if(this.inProgress()){
     return this.nextLevelBluePrints.time - ( this.remainingBuildTime - Math.ceil((new Date().getTime() - this._LoadTime)/1000) );
    }else{
      return 0;
    }
  },
  
  textInfo : function(){
    return this.humanizeString(this.name)+ "<br/>Level " + this.level;
  },
	
  humanizeString : function(str){
  	return str.capitalize().replace("_", " ") 
  },
  
  getMeterFunc: function(){
  	var self = this
	  return function(){
	    	return self.hp/self.maxHp	
	  }
  },
  
  isValidToUpgrade : function(silent){
    if(this.game.disableJsValidation)
      return true;
    /****************************** Validating workers **************************************/
    if( this.game.workerFactory.idleWorkers == 0 ){
      if(!silent) Notification.alert("Cannot upgrade " + this.name + ", all your workers are busy!");
      return false;
    }
    /****************************** Validating upgrade **************************************/    
    var level = (parseInt(this.level) + 1).toString(); 
    if(!this.factory.bluePrints.levels[level]) {
      if(!silent) Notification.alert("You have reached max upgrade");
      return false;
    }
    /******************************************************************************************/
    /****************************** Validating resources **************************************/
    var level = (parseInt(this.level) + 1).toString(); 
    var neededRock = this.factory.bluePrints.levels[level].rock - this.game.resources.rock;
    var neededLumber = this.factory.bluePrints.levels[level].lumber - this.game.resources.lumber;
    
    if( neededRock > 0 && neededLumber > 0 ){
      if(!silent) Notification.alert("Not enough resources, you need more "+ neededRock +" rock and "+ neededLumber + " lumber");
      return false;
    }
    
    if(  neededRock > 0 ){
      if(!silent) Notification.alert("Not enough resources, you need more "+ neededRock +" rock");
      return false;
    }
    
    if( neededLumber > 0 ){
      if(!silent) Notification.alert("Not enough resources, you need more "+ neededLumber +" lumber");
      return false;
    }
    /****************************** Validating dependencies *********************************/
    var result = this.factory.isDependenciesMet(this.level+1);
    return result.valid
  },

  isValidToBuild : function(x,y) {
    if(this.game.disableJsValidation)
      return true;
    /****************************** Validating dependencies *********************************/
    var result = this.factory.isDependenciesMet(this.level+1);
    /*************** Validating max number of buildings, workers, resources *****************/
    if(result.valid)
      result = this.factory.hasEnoughResources(this.level+1);
    else
      return result.valid;
    /****************************** Validate Location ***************************************/
    if(result.valid && this.validateLocation(x,y)){
			Map.addElement(this)
		}
		else{
			return false;
		} 
    return true;
  },
  getUpgradeSpecs: function(){
  	var upgradeSpecs = {
  		level: 0,
  		rock: 0,
  		wood: 0,
  		gold: 0,
  		time: 0,
  		requirements: {},
  		gains: {},
  		maxUpgrade: false,
  		workersNotEnough: false,
  	}
  	if (this.game.workerFactory.idleWorkers == 0) {
  		upgradeSpecs['workersNotEnough'] = true
  	}
  	
  	var level = (parseInt(this.level) + 1).toString();
  	if (!this.factory.bluePrints.levels[level]) {
  		upgradeSpecs['maxUpgrade'] = true
  		return upgradeSpecs
  	}
  	upgradeSpecs['level'] = level
	upgradeSpecs['rock'] = {}
	upgradeSpecs['lumber'] = {}
  	upgradeSpecs['rock'].value = this.factory.bluePrints.levels[level].rock
	upgradeSpecs['lumber'].value = this.factory.bluePrints.levels[level].lumber
	upgradeSpecs['notEnoughResources'] = false
	upgradeSpecs['dependenciesInvalid'] = false
	if(this.game.resources.rock>=upgradeSpecs['rock'].value){
		upgradeSpecs['rock'].valid = true	
	}else{
		upgradeSpecs['notEnoughResources'] = true
		upgradeSpecs['rock'].valid = false
	}
	if(this.game.resources.lumber>=upgradeSpecs['lumber'].value){
		upgradeSpecs['lumber'].valid = true
	}else{
		upgradeSpecs['notEnoughResources'] = true
		upgradeSpecs['lumber'].valid = false
	}	 
  	
	upgradeSpecs['time'] = Util.timeDisplay(this.nextLevelBluePrints.time).gsub('seconds','s').gsub(' ','')
	upgradeSpecs['requirements'] = this.factory.getDependenciesValidations(level)
	upgradeSpecs['requirements'].each(function(requirement){
		if(!requirement.valid){
			upgradeSpecs['dependenciesInvalid'] = true
		}
	})
  	return upgradeSpecs
  },
  getUpgradableSpecs : function(){
    var upgradableSpecs = {}
    for(property in this.currentLevelBluePrints){
      if(currentLevelBluePrints[property].upgradable){
        upgradableSpecs[property]= {}
        upgradableSpecs[property]['from'] = this.currentLevelBluePrints[property]
        upgradableSpecs[property]['to'] = this.nextLevelBluePrints[property] 
      }
    }
    return upgradableSpecs
  }
});
